File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.*;
import java.util.*;

public class Main {

    static int v;
    static int[] p = new int[10002];

    public static int find(int x) {
        if (p[x] < 0) return x;
        return p[x] = find(p[x]);
    }

    public static boolean union(int x, int y) {
        x = find(x); y = find(y);
        if (x == y) return false;
        
        p[x] = y;
        return true;
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        v = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());
        Arrays.fill(p, 0, v + 1, -1);

        // 0 : u, 1 : v, 2 : cost
        PriorityQueue<int[]> pq = new PriorityQueue<>((int[] a, int[] b) -> a[2] - b[2]);
        while (e-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int cost = Integer.parseInt(st.nextToken());
            pq.offer(new int[]{u, v, cost});
        }

        long result = 0;
        int cnt = 0;
        while (!pq.isEmpty()) {
            var cur = pq.poll();
            int p = cur[0];
            int q = cur[1];
            int cost = cur[2];

            if (!union(p, q)) continue;
            
            result += cost;
            cnt++;
            if (cnt == v - 1) break;
        }

        System.out.print(result);
    }
}

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.kt
import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val (v, e) = with(StringTokenizer(readLine())) {
        listOf(nextToken().toInt(), nextToken().toInt())
    }

    val edges = mutableListOf<Triple<Int, Int, Int>>()
    repeat(e) {
        with(StringTokenizer(readLine())) {
            edges.add(Triple(nextToken().toInt(), nextToken().toInt(), nextToken().toInt()))
        }
    }
    edges.sortBy { it.third }

    val p = IntArray(v + 1) { -1 }
    fun find(x: Int): Int {
        if (p[x] < 0) return x
        p[x] = find(p[x])
        return p[x]
    }

    fun union(m: Int, n: Int): Boolean {
        var x = find(m)
        var y = find(n)
        if (x == y) return false

        if (x > y) x = y.also { y = x }
        if (p[x] == p[y]) p[x]--

        p[y] = x
        return true
    }

    var cnt = 0
    var result = 0
    for (edge in edges) {
        if (cnt == v - 1) break

        if (union(edge.first, edge.second)) {
            result += edge.third
            cnt++
        }
    }

    print(result)
}
